tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -11539,6 +11592,66 @@ void SpirvEmitter::processMeshOutputCounts(const CallExpr *callExpr) {
   }
 }
 
+SpirvInstruction*
+SpirvEmitter::processGetAttributeAtVertex(const CallExpr *expr) {
+  const auto exprLoc = expr->getExprLoc();
+  const auto exprRange = expr->getSourceRange();
+  QualType arg0Type;
+
+  // arg1 : vertexId
+  const auto *arg1BaseExpr = doExpr(expr->getArg(1)->IgnoreParenLValueCasts());
+  const auto *arg1ConstExpr = dyn_cast<SpirvConstantInteger>(arg1BaseExpr);
+  const auto *vertexId = arg1ConstExpr->getValue().getRawData();
+
+  // arg0 : <NoInterpolation> decorated input
+  // Tip  : for input with boolean type, we need to ignore implicit cast first,
+  //        to match surrounded workaround cast expr.
+  auto *arg0NoCast = expr->getArg(0)->IgnoreCasts();
+  bool hasImplicitCast = false;
+
+  if (dyn_cast<CastExpr>(expr->getArg(0)->IgnoreParenLValueCasts())) {
+    // If stage inputs is boolean type, it has been cast to uint type.
+    hasImplicitCast = true;
+  }
+
+  if (dyn_cast<MemberExpr>(arg0NoCast)) {
+    // As a structure field
+    const auto *arg0NamedDecl = (dyn_cast<MemberExpr>(arg0NoCast))->getFoundDecl().getDecl();
+    arg0Type = dyn_cast<ValueDecl>(arg0NamedDecl)->getType()->getAsArrayTypeUnsafe()->getElementType();
```

Comment:
segmentation fault if the nointerpolation attribute is used on the struct:
```hlsl
enum VertexID {
    FIRST = 0,
    SECOND = 1,
    THIRD = 2
};

struct PSInput
{
    //float4 position : SV_POSITION;
    float3 color1 : COLOR1;
    bool3 color2 : COLOR2;
};

float3 main(nointerpolation PSInput input) : SV_Target
{
    float3 vColor0 = GetAttributeAtVertex( input.color1, VertexID::SECOND );
    float3 vColor1 = GetAttributeAtVertex( input.color2, VertexID::THIRD );
    return (vColor0 + vColor1);
}
```

Reply:
Same crash if I remove the nointerpolation altogether:
```hlsl
enum VertexID {
    FIRST = 0,
    SECOND = 1,
    THIRD = 2
};

struct PSInput
{
    float3 color1 : COLOR1;
    bool3 color2 : COLOR2;
};

float3 main(float4 position : SV_POSITION, PSInput input) : SV_Target
{
    float3 vColor0 = GetAttributeAtVertex(input.color1, VertexID::SECOND );
    float3 vColor1 = GetAttributeAtVertex(input.color2, VertexID::THIRD );
    return (vColor0 + vColor1);
}
```

For the nointerpolation on the struct, I believe it should be valid?
For the second case, seems like there should be an error as I pass an attribute which is not marked as "nointerpolation" no?

Reply:
For case 1, I think I miss this case, would add a new commit for decorated struct.
For case 2, yes, need to report error message for this.

Reply:
Case 1 : supported.
Case 2 : errors could be observed when running dxc:

```
enum VertexID { 
    FIRST = 0,
    SECOND = 1,
    THIRD = 2
};

struct PSInput
{
    float4 position : SV_POSITION;
    float3 color1 : COLOR1;
    bool3 color2 : COLOR2;
};

float3 main(PSInput input,
            bool3 color2 : COLOR2) : SV_Target
{
    float3 vColor0 = GetAttributeAtVertex( input.color1, VertexID::SECOND );
    float3 vColor1 = GetAttributeAtVertex( color2, VertexID::SECOND );
    return vColor0 + vColor0;
}
12:12: error: First parameter of GetAttributeAtVertex should be decorated with 'nointerpolation'.
17:19: error: First parameter of GetAttributeAtVertex should be decorated with 'nointerpolation'.
```